diff -up opencryptoki-3.11.0/usr/sbin/pkcsconf/pkcsconf.c.me opencryptoki-3.11.0/usr/sbin/pkcsconf/pkcsconf.c
--- opencryptoki-3.11.0/usr/sbin/pkcsconf/pkcsconf.c.me	2019-01-31 13:27:05.720647942 +0100
+++ opencryptoki-3.11.0/usr/sbin/pkcsconf/pkcsconf.c	2019-01-31 13:36:05.763624633 +0100
@@ -1072,6 +1072,9 @@ CK_RV init(void)
     if (rc != CKR_OK) {
         printf("Error initializing the PKCS11 library: 0x%lX (%s)\n", rc,
                p11_get_ckr(rc));
+        printf("Note: all non-root users that require access to PKCS#11 tokens "
+               "using opencryptoki must be assigned to the pkcs11 group to be "
+               "able to communicate with the pkcsslotd daemon.\n");
         fflush(stdout);
         cleanup();
     }
