diff -up opencryptoki-3.10.0/usr/lib/pkcs11/common/mech_ec.c.me opencryptoki-3.10.0/usr/lib/pkcs11/common/mech_ec.c
--- opencryptoki-3.10.0/usr/lib/pkcs11/common/mech_ec.c.me	2018-06-06 21:55:55.000000000 +0200
+++ opencryptoki-3.10.0/usr/lib/pkcs11/common/mech_ec.c	2018-10-26 13:52:11.461448596 +0200
@@ -797,6 +797,7 @@ ckm_kdf_X9_63(STDLL_TokData_t *tokdata,
         counter++;
     }
 
+    free(ctx);
     return CKR_OK;
 }
 
diff -up opencryptoki-3.10.0/usr/lib/pkcs11/common/mech_ssl3.c.me opencryptoki-3.10.0/usr/lib/pkcs11/common/mech_ssl3.c
--- opencryptoki-3.10.0/usr/lib/pkcs11/common/mech_ssl3.c.me	2018-06-06 21:55:55.000000000 +0200
+++ opencryptoki-3.10.0/usr/lib/pkcs11/common/mech_ssl3.c	2018-10-26 13:51:12.538255825 +0200
@@ -1195,7 +1195,7 @@ error:
    if (value_len_attr)    free( value_len_attr );
    if (always_sens_attr)  free( always_sens_attr );
    if (extract_attr)      free( extract_attr );
-
+   if (derived_key_obj)   object_free(derived_key_obj);
    return rc;
 }
 
