diff -up opencryptoki-3.10.0/usr/lib/pkcs11/ep11_stdll/ep11_specific.c.me opencryptoki-3.10.0/usr/lib/pkcs11/ep11_stdll/ep11_specific.c
--- opencryptoki-3.10.0/usr/lib/pkcs11/ep11_stdll/ep11_specific.c.me	2018-12-14 18:48:48.945096527 +0100
+++ opencryptoki-3.10.0/usr/lib/pkcs11/ep11_stdll/ep11_specific.c	2018-12-14 18:55:18.198260044 +0100
@@ -6268,7 +6268,7 @@ CK_RV ep11tok_login_session(STDLL_TokDat
         }
     }
 
-    rc = handle_all_ep11_cards((ep11_target_t *)&ep11_data->target_list,
+    rc = handle_all_ep11_cards((ep11_target_t *)ep11_data->target_list,
                                ep11_login_handler, ep11_session);
     if (rc != CKR_OK) {
         TRACE_ERROR("%s handle_all_ep11_cards failed: 0x%lu\n", __func__, rc);
@@ -6310,7 +6310,7 @@ CK_RV ep11tok_login_session(STDLL_TokDat
 done:
     if (rc != CKR_OK) {
         if (ep11_session->flags & (EP11_SESS_PINBLOB_VALID | EP11_VHSM_PINBLOB_VALID)) {
-            rc2 = handle_all_ep11_cards((ep11_target_t *)&ep11_data->target_list,
+            rc2 = handle_all_ep11_cards((ep11_target_t *)ep11_data->target_list,
                                         ep11_logout_handler, ep11_session);
             if (rc2 != CKR_OK)
                 TRACE_ERROR("%s handle_all_ep11_cards failed: 0x%lu\n", __func__, rc2);
@@ -6356,7 +6356,7 @@ CK_RV ep11tok_relogin_session(STDLL_TokD
         return CKR_USER_NOT_LOGGED_IN;
     }
 
-    rc = handle_all_ep11_cards((ep11_target_t *)&ep11_data->target_list,
+    rc = handle_all_ep11_cards((ep11_target_t *)ep11_data->target_list,
                                ep11_login_handler, ep11_session);
     if (rc != CKR_OK)
         TRACE_ERROR("%s handle_all_ep11_cards failed: 0x%lu\n", __func__, rc);
@@ -6401,7 +6401,7 @@ CK_RV ep11tok_logout_session(STDLL_TokDa
         return CKR_USER_NOT_LOGGED_IN;
     }
 
-    rc = handle_all_ep11_cards((ep11_target_t *)&ep11_data->target_list,
+    rc = handle_all_ep11_cards((ep11_target_t *)ep11_data->target_list,
                                ep11_logout_handler, ep11_session);
     if (rc != CKR_OK)
         TRACE_ERROR("%s handle_all_ep11_cards failed: 0x%lu\n", __func__, rc);
