--- a/configure.ac	2020-06-21 10:55:26.415866708 +0200
+++ b/configure.ac	2020-06-21 11:02:22.234644303 +0200
@@ -4080,7 +4080,7 @@
 #include <stdatomic.h>
 ],
 [
-  #if ATOMIC_INT_LOCK_FREE == 2 && ATOMIC_LONG_LOCK_FREE == 2
+  #if ATOMIC_INT_LOCK_FREE >= 1 && ATOMIC_LONG_LOCK_FREE >= 1
   atomic_int_fast32_t val = 0; atomic_fetch_add_explicit(&val, 1, memory_order_relaxed);
   #else
   #error stdatomic not lock free
