diff -up mesa-19.3.3/src/glx/drisw_glx.c.dma mesa-19.3.3/src/glx/drisw_glx.c
--- mesa-19.3.3/src/glx/drisw_glx.c.dma	2020-02-14 12:37:42.551008273 +1000
+++ mesa-19.3.3/src/glx/drisw_glx.c	2020-02-14 12:37:49.683081406 +1000
@@ -199,7 +199,7 @@ swrastXPutImage(__DRIdrawable * draw, in
       XShmPutImage(dpy, drawable, gc, ximage, srcx, srcy, x, y, w, h, False);
       XSync(dpy, False);
    } else {
-      ximage->width = w;
+      ximage->width = ximage->bytes_per_line / ((ximage->bits_per_pixel + 7)/ 8);
       ximage->height = h;
       XPutImage(dpy, drawable, gc, ximage, srcx, srcy, x, y, w, h);
    }
