From 4822b2805d70863c96445476e86eb7630fbc1f37 Mon Sep 17 00:00:00 2001
From: Jacco Ligthart <jacco@redsleeve.org>
Date: Sat, 23 Nov 2019 22:49:44 +0100
Subject: [PATCH] added arm to the ExclusiveArch and nobuildarches lists

---
 SPECS/kernel.spec | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/SPECS/kernel.spec b/SPECS/kernel.spec
index 9d90f97..c4aad94 100644
--- a/SPECS/kernel.spec
+++ b/SPECS/kernel.spec
@@ -247,6 +247,11 @@
 %define kernel_image arch/arm64/boot/Image.gz
 %endif
 
+%ifarch %{arm}
+%define asmarch arm
+%define hdrarch arm
+%endif
+
 # To temporarily exclude an architecture from being built, add it to
 # %%nobuildarches. Do _NOT_ use the ExclusiveArch: line, because if we
 # don't build kernel-headers then the new build system will no longer let
@@ -254,7 +259,7 @@
 # Which is a BadThing(tm).
 
 # We only build kernel-headers on the following...
-%define nobuildarches i386 i686
+%define nobuildarches i386 i686 %{arm}
 
 %ifarch %nobuildarches
 %define with_up 0
@@ -284,11 +289,11 @@ Group: System Environment/Kernel
 License: GPLv2 and Redistributable, no modification permitted
 URL: http://www.kernel.org/
 Version: %{rpmversion}
-Release: %{pkg_release}
+Release: %{pkg_release}.redsleeve
 Summary: The Linux kernel, based on version %{version}, heavily modified with backports
 # DO NOT CHANGE THE 'ExclusiveArch' LINE TO TEMPORARILY EXCLUDE AN ARCHITECTURE BUILD.
 # SET %%nobuildarches (ABOVE) INSTEAD
-ExclusiveArch: noarch i386 i686 x86_64 s390x aarch64 ppc64le
+ExclusiveArch: noarch i386 i686 x86_64 s390x aarch64 ppc64le %{arm}
 ExclusiveOS: Linux
 %ifnarch %{nobuildarches}
 Requires: kernel-core-uname-r = %{KVERREL}%{?variant}
@@ -2317,6 +2322,9 @@ fi
 #
 #
 %changelog
+* Sat Nov 23 2019 Jacco Ligthart <jacco@redsleeve.org> 4.18.0-147.0.3.el8.redsleeve
+- added arm to the ExclusiveArch and nobuildarches lists
+
 * Wed Nov 13 2019 CentOS Sources <bugs@centos.org> - 4.18.0-147.0.3.el8.centos
 - Apply debranding changes
 
-- 
2.18.1

