diff --git a/kernel/power/drot.c b/kernel/power/drot.c
index 3e10748..9d2b8c0 100644
--- a/kernel/power/drot.c
+++ b/kernel/power/drot.c
@@ -37,8 +37,6 @@ USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #include "common.h"
 
-#pragma GCC optimize "O1"
-
 #if defined(POWER8)
 #include "drot_microk_power8.c"
 #endif
diff --git a/kernel/power/srot.c b/kernel/power/srot.c
index d2910ff..18f5362 100644
--- a/kernel/power/srot.c
+++ b/kernel/power/srot.c
@@ -37,8 +37,6 @@ USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #include "common.h"
 
-#pragma GCC optimize "O1"
-
 #if defined(POWER8)
 #include "srot_microk_power8.c"
 #endif
diff --git a/kernel/power/zscal.c b/kernel/power/zscal.c
index 14d677f..79f7271 100644
--- a/kernel/power/zscal.c
+++ b/kernel/power/zscal.c
@@ -36,8 +36,6 @@ USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #include "common.h"
 
-#pragma GCC optimize "O1"
-
 #if defined(POWER8)
 #if defined(DOUBLE)
 #include "zscal_microk_power8.c"
