diff --git a/kernel/zarch/izamax.c b/kernel/zarch/izamax.c
index 216c341..6cde691 100644
--- a/kernel/zarch/izamax.c
+++ b/kernel/zarch/izamax.c
@@ -185,7 +185,7 @@ static BLASLONG   ziamax_kernel_16_TUNED(BLASLONG n, FLOAT *x, FLOAT *maxf) {
             "vsteg  %%v6,%[maxf],0  \n\t"
             "vmnlg  %%v1,%%v5,%%v7 \n\t"
             "vlgvg  %[index],%%v1,0  \n\t"
-            "j 3    \n\t"
+            "j 3f   \n\t"
             "2:     \n\t"
             "wfchdb %%v16,%%v26,%%v6      \n\t"
             "vsel   %%v1,%%v5,%%v7,%%v16  \n\t"
