From 5eed36bd2a508211ed30efd10f635b15f6cd6be3 Mon Sep 17 00:00:00 2001
From: Jacco Ligthart <jacco@redsleeve.org>
Date: Wed, 12 Feb 2020 21:50:15 +0100
Subject: [PATCH] disabled valgrind for arm

---
 SPECS/python2.spec | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/SPECS/python2.spec b/SPECS/python2.spec
index 1c3cc9a..f963209 100644
--- a/SPECS/python2.spec
+++ b/SPECS/python2.spec
@@ -48,7 +48,7 @@
 %global with_systemtap 1
 
 # some arches don't have valgrind so we need to disable its support on them
-%ifnarch s390 %{mips} riscv64
+%ifnarch s390 %{mips} riscv64 %{arm}
 %global with_valgrind 1
 %else
 %global with_valgrind 0
@@ -104,7 +104,7 @@ Summary: An interpreted, interactive, object-oriented programming language
 Name: %{python}
 # Remember to also rebase python2-docs when changing this:
 Version: 2.7.16
-Release: 12%{?dist}
+Release: 12.redsleeve%{?dist}
 License: Python
 Group: Development/Languages
 Requires: %{python}-libs%{?_isa} = %{version}-%{release}
@@ -2030,6 +2030,9 @@ fi
 # ======================================================
 
 %changelog
+* Wed Nov 13 2019 Jacco Ligthart <jacco@redsleeve.org> - 2.7.16-12.redsleeve
+- disabled valgrind for arm
+
 * Tue Sep 03 2019 Tomas Orsava <torsava@redhat.com> - 2.7.16-12
 - Adding FIPS compliance to Python 2 in RHEL8:
   - Updated patch 146 with a new version of the FIPS patch
-- 
2.18.1

