From 43294f6ac6de3b83badbfe118348d222cbe7a32d Mon Sep 17 00:00:00 2001
From: Jacco Ligthart <jacco@redsleeve.org>
Date: Sat, 5 Dec 2020 00:21:59 +0100
Subject: [PATCH] Don't exclude arm architectures

---
 SPECS/numactl.spec | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/SPECS/numactl.spec b/SPECS/numactl.spec
index 93be941..39e5124 100644
--- a/SPECS/numactl.spec
+++ b/SPECS/numactl.spec
@@ -1,7 +1,7 @@
 Name:		numactl
 Summary:	Library for tuning for Non Uniform Memory Access machines
 Version:	2.0.12
-Release:	11%{?dist}
+Release:	11%{?dist}.redsleeve
 # libnuma is LGPLv2 and GPLv2
 # numactl binaries are GPLv2 only
 License:	GPLv2
@@ -11,7 +11,7 @@ Source0:	https://github.com/numactl/numactl/releases/download/%{version}/numactl
 Buildroot:	%{_tmppath}/%{name}-buildroot
 BuildRequires:	libtool automake autoconf
 
-ExcludeArch: s390 %{arm}
+ExcludeArch: s390
 
 #START INSERT
 #
@@ -136,6 +136,9 @@ make DESTDIR=$RPM_BUILD_ROOT install
 %{_mandir}/man3/*.3*
 
 %changelog
+* Fri Dec 04 2020 Jacco Ligthart <jacco@redsleeve.org> - 2.0.12-11.redsleeve
+- Don't exclude arm architectures
+
 * Sat May  9 2020 Pingfan Liu <piliu@redhat.com> - 2.0.12-11
 - Update manpage description of --localalloc option
 
-- 
2.18.4

