diff --git a/src/qdoc/configure.pri b/src/qdoc/configure.pri
index 17008c3..8a6f01a 100644
--- a/src/qdoc/configure.pri
+++ b/src/qdoc/configure.pri
@@ -109,7 +109,7 @@ defineTest(qtConfTest_libclang) {
         output = $$system("$$llvm_config --version 2>/dev/null")
         CLANG_VERSION = $$extractVersion($$output)
     } else {
-        CLANG_LIBDIR = $$clangInstallDir/lib
+        CLANG_LIBDIR = $$clangInstallDir/lib64
         CLANG_INCLUDEPATH = $$clangInstallDir/include
         CLANG_VERSION = $$findLLVMVersionFromLibDir($$CLANG_LIBDIR)
     }
