diff --git a/mkspecs/features/qt_find_clang.prf b/mkspecs/features/qt_find_clang.prf
index 583cf0e..361ba85 100644
--- a/mkspecs/features/qt_find_clang.prf
+++ b/mkspecs/features/qt_find_clang.prf
@@ -72,7 +72,7 @@ for(_, $$list(_)) { # just a way to break easily
         output = $$system("$$llvm_config --version 2>/dev/null")
         CLANG_VERSION = $$extractVersion($$output)
     } else {
-        CLANG_LIBDIR = $$clangInstallDir/lib
+        CLANG_LIBDIR = $$clangInstallDir/lib64
         CLANG_INCLUDEPATH = $$clangInstallDir/include
         CLANG_VERSION = $$findLLVMVersionFromLibDir($$CLANG_LIBDIR)
     }
