From 08f6497cd0eb36d72a117b8fb8e8c48281eb7775 Mon Sep 17 00:00:00 2001
From: Jacco Ligthart <jacco@redsleeve.org>
Date: Sat, 8 Feb 2020 22:40:25 +0100
Subject: [PATCH] added %{arm} to the no hardware architectures

---
 SPECS/mesa.spec | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/SPECS/mesa.spec b/SPECS/mesa.spec
index 472b83f..d4bdb22 100644
--- a/SPECS/mesa.spec
+++ b/SPECS/mesa.spec
@@ -1,7 +1,7 @@
 %global llvm_toolset %{nil}
 %global llvm_pkg_prefix %{nil}
 
-%ifarch s390x
+%ifarch s390x %{arm}
 %define with_hardware 0
 %else
 %define with_hardware 1
@@ -20,7 +20,7 @@
 %define with_vulkan 0
 %endif
 
-%ifarch %{arm} aarch64
+%ifarch aarch64
 %define with_xa        1
 %endif
 
@@ -41,7 +41,7 @@
 Name:           mesa
 Summary:        Mesa graphics libraries
 Version:        19.1.4
-Release:        3%{?rctag:.%{rctag}}%{?dist}
+Release:        3%{?rctag:.%{rctag}}%{?dist}.redsleeve
 
 License:        MIT
 URL:            http://www.mesa3d.org
@@ -602,6 +602,9 @@ done
 %endif
 
 %changelog
+* Fri Feb 07 2020 Jacco Ligthart <jacco@redsleeve.org> - 19.1.4-3.redsleeve
+- added %{arm} to the no hardware architectures
+
 * Mon Nov 25 2019 Ben Crocker <bcrocker@redhat.com> - 19.1.4-3
 - Patch to require Large CodeModel for llvmpipe on ppc64
 
-- 
2.18.1

