From 0710c9bd00dbebc3fcd67d511e903684cf179d93 Mon Sep 17 00:00:00 2001
From: Jacco Ligthart <jacco@redsleeve.org>
Date: Sat, 4 Dec 2021 20:26:38 +0000
Subject: [PATCH] removed arm from exclude archs

---
 SPECS/opensm.spec | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/SPECS/opensm.spec b/SPECS/opensm.spec
index fb5787d..a18a570 100644
--- a/SPECS/opensm.spec
+++ b/SPECS/opensm.spec
@@ -2,7 +2,7 @@
 
 Name:    opensm
 Version: 3.3.24
-Release: 1%{?dist}
+Release: 1%{?dist}.redsleeve
 Summary: OpenIB InfiniBand Subnet Manager and management utilities
 Group:   System Environment/Daemons
 License: GPLv2 or BSD
@@ -26,7 +26,7 @@ Requires(post): systemd
 Requires(preun): systemd
 Requires(postun): systemd
 # RDMA is not currently built on 32-bit ARM: #1484155
-ExcludeArch: s390 %{arm}
+ExcludeArch: s390
 
 %description
 OpenSM is the OpenIB project's Subnet Manager for Infiniband networks.
@@ -125,6 +125,9 @@ fi
 %{_libdir}/lib*.a
 
 %changelog
+* Sat Dec 04 2021 Jacco Ligthart <jacco@redsleeve.org> - 3.3.24-1.redsleeve
+- removed arm from exclude archs
+
 * Tue Jun 15 2021 Honggang Li <honli@redhat.com> - 3.3.24-1
 - Rebase to latest upstream release 3.3.24
 - Resolves: bz1961961
-- 
2.27.0

